### PR TITLE
Refactor Examples

### LINKED ISSUE #810: [Docs] Proper usage of ScaleKernel in examples
# 📚 Documentation/Examples
See #716

### LINKED ISSUE #914: [Example] create example notebook for GPs with uncertain inputs
Based on #659 and #913, we probably want some examples that show how to train GPs with uncertain inputs.

1) Inputs with known uncertainty - can be trained using SVGP, where we sample from the distribution of inputs

2) Unknown inputs that we want to impute - should be done using Pyro

### LINKED ISSUE #945: [Bug] Example for Deep Kernel Learning is out of date.
# 🐛 Bug

When running the CIFAR 10 example, it gives a shape mismatch error.

## To reproduce

Run this line by line in iPython (use dataset = 'cifar10' as in the example):

https://github.com/cornellius-gp/gpytorch/blob/master/examples/08_Deep_Kernel_Learning/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.ipynb

```

** Stack trace/error message **

     49     def initialize_variational_distribution(self, prior_dist):
---> 50         self.variational_mean.data.copy_(prior_dist.mean)
     51         self.variational_mean.data.add_(self.mean_init_std, torch.randn_like(prior_dist.mean))
     52         self.chol_variational_covar.data.copy_(prior_dist.lazy_covariance_matrix.cholesky().evaluate())

RuntimeError: output with shape [64] doesn't match the broadcast shape [132, 64]
 ```
from
```
gpytorch/variational/cholesky_variational_distribution.py in initialize_variational_distribution(self, prior_dist)
``` 

## Expected Behavior

It should train a DKL, but errors when training.

## System information

**Please complete the following information:**
-  0.3.6
-  1.3.1
-  Ubuntu 16.04.6

### LINKED ISSUE #948: [Docs] Add Multitask variational GP example.
Based on an issue in #933.

### LINKED ISSUE #961: [Bug] Pyro Examples are out of date
# 🐛 Bug

I'm attempting to run the Pyro examples in [examples/09_Pyro_Integration](https://github.com/cornellius-gp/gpytorch/tree/master/examples/09_Pyro_Integration), but am getting backwards related issues.

## To reproduce

Run the [Pyro GP Regression](https://github.com/cornellius-gp/gpytorch/blob/master/examples/09_Pyro_Integration/Pyro_Variational_GP_Regression.ipynb) tutorial into cell [5] and gives the following error. However, I've seen it through all three tutorials. 

** Code snippet to reproduce **

Error seems to require clearing a gradient step in the iteration perspective, but I'm not particularly familiar enough with Pyro to come up with a fix off the top of my head.

** Stack trace/error message **
```
<ipython-input-5-b72bfb4618e0> in train(num_iter)
      9     for i in range(num_iter):
     10         model.zero_grad()
---> 11         loss = svi.step(train_x, train_y)
     12         if not (i + 1) % 50:
     13             print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f   noise: %.3f' % (

~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/pyro/infer/svi.py in step(self, *args, **kwargs)
    117         # get loss and compute gradients
    118         with poutine.trace(param_only=True) as param_capture:
--> 119             loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)
    120 
    121         params = set(site["value"].unconstrained()

~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/pyro/infer/trace_elbo.py in loss_and_grads(self, model, guide, *args, **kwargs)
    132             if trainable_params and getattr(surrogate_loss_particle, 'requires_grad', False):
    133                 surrogate_loss_particle = surrogate_loss_particle / self.num_particles
--> 134                 surrogate_loss_particle.backward(retain_graph=self.retain_graph)
    135         warn_if_nan(loss, "loss")
    136         return loss

~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/tensor.py in backward(self, gradient, retain_graph, create_graph)
    164                 products. Defaults to ``False``.
    165         """
--> 166         torch.autograd.backward(self, gradient, retain_graph, create_graph)
    167 
    168     def register_hook(self, hook):

~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
     97     Variable._execution_engine.run_backward(
     98         tensors, grad_tensors, retain_graph, create_graph,
---> 99         allow_unreachable=True)  # allow_unreachable flag
    100 
    101 

RuntimeError: Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.
```

## Expected Behavior

The tutorial should work :). Sorry if this is just reporting a bug. But, it should be part of the examples refactor.

## System information

**Please complete the following information:**
- Pyro 1.0.0 
- gpytorch 0.3.6 (built from master) 
- pytorch 1.3.1 
- Ubuntu 18.04

### LINKED ISSUE #995: [Bug] DKL example fails with broadcast error
# 🐛 Bug

When I run Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial, it fails with the following error:
RuntimeError: output with shape [64] doesn't match the broadcast shape [132, 64].

## To reproduce
Run Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.ipynb from the examples folder - last cell will crash.

** Stack trace/error message **
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-6f7e8ea00fe7> in <module>
      2     scheduler.step()
      3     with gpytorch.settings.use_toeplitz(False):
----> 4         train(epoch)
      5         test()
      6     state_dict = model.state_dict()

<ipython-input-8-2fe6607c5f10> in train(epoch)
     19         data, target = data.cuda(), target.cuda()
     20         optimizer.zero_grad()
---> 21         output = model(data)
     22         print(output.shape, target.shape)
     23         loss = -mll(output, target)

~/anaconda3/envs/lapki/lib/python3.7/site-packages/gpytorch/module.py in __call__(self, *inputs, **kwargs)
     22 
     23     def __call__(self, *inputs, **kwargs):
---> 24         outputs = self.forward(*inputs, **kwargs)
     25         if isinstance(outputs, list):
     26             return [_validate_module_outputs(output) for output in outputs]

<ipython-input-7-834e655c1eac> in forward(self, x)
     10         features = self.feature_extractor(x)
     11         features = gpytorch.utils.grid.scale_to_bounds(features, self.grid_bounds[0], self.grid_bounds[1])
---> 12         res = self.gp_layer(features)
     13         return res
     14 

~/anaconda3/envs/lapki/lib/python3.7/site-packages/gpytorch/models/approximate_gp.py in __call__(self, inputs, prior, **kwargs)
     54         if inputs.dim() == 1:
     55             inputs = inputs.unsqueeze(-1)
---> 56         return self.variational_strategy(inputs, prior=prior)

~/anaconda3/envs/lapki/lib/python3.7/site-packages/gpytorch/variational/_variational_strategy.py in __call__(self, x, prior)
    106         if not self.variational_params_initialized.item():
    107             prior_dist = self.prior_distribution
--> 108             self._variational_distribution.initialize_variational_distribution(prior_dist)
    109             self.variational_params_initialized.fill_(1)
    110 

~/anaconda3/envs/lapki/lib/python3.7/site-packages/gpytorch/variational/cholesky_variational_distribution.py in initialize_variational_distribution(self, prior_dist)
     48 
     49     def initialize_variational_distribution(self, prior_dist):
---> 50         self.variational_mean.data.copy_(prior_dist.mean)
     51         self.variational_mean.data.add_(self.mean_init_std, torch.randn_like(prior_dist.mean))
     52         self.chol_variational_covar.data.copy_(prior_dist.lazy_covariance_matrix.cholesky().evaluate())

RuntimeError: output with shape [64] doesn't match the broadcast shape [132, 64]
```

## Expected Behavior

I have expected this to train.

## System information

**Please complete the following information:**
-  GPytorch:  0.3.6    
-  Pytorch: 1.3.1   
-  OS: Ubuntu 19.10, 5.3.0-24;   
-  Python: 3.7.5

### PR DESCRIPTION
**Pull request by gpleiss**
Looking for input from @jacobrgardner and @Balandat 

I think our examples have gotten out of hand a bit and needed to be re-organized and consolidated.  The current structure I have is:

00_Basic_Usage
01_Exact_GPs (contains simple regression, spectral mixture regression, NUTS example)
02_Scalable_Exact_GPs (contains (multi) GPU examples, KeOPS, Grid GP, SVGP, KISS-GP, SKIP)
03_Multitask_Exact_GPs (contains kronecker/hadamard, model list)
04_Variational_and_Approximate_GPs (contains SVGP, Predictive log likelihood, GP classification. Will also include imputing data and multi-output SVGP).
05_Deep_Gaussian_Processes
06_Deep_Kernel_Learning (contains KISSGP and CIFAR examples)
07_Advanced_Usage (contains batch mode, derivatives)
08_Pyro_Integration (to be refactored)

- [x] Add example smoke testing in CI

New examples:
- [x] Batch independent multitask GPs
- [x] Missing data
- [x] Multitask variational example
- [x] Pyro multitask cluster
- [x] Pyro cox process

Examples to fix:
- [x] Pyro integration
- [x] Pyro latent function
- [x] DKL example
- [x] SVGP example
- [x] Classification example
- [x] Deep GP?
- [x] CIFAR10 tutorial

Extra updates:
- [x] Fix error in `ndimension` for LazyEvaluatedKernelTensor
- [x] Update `SoftmaxLikelihood` to use new multitask interface
- [x] Fix bugs in `GridInterpolationKernel`, pyro `Likelihood` interface, caching issues with `MultitaskVariationalStrategy`.

Closes #961 
Closes #948 
Closes #945 
Closes #914 
Closes #810 
Closes #995

### CODE REVIEW

**Comment by jacobrgardner** [comment]
The structure definitely looks good to me.

One suggestion, what if we rephrase 06 as "Deep Kernel Learning and PyTorch Integration" (or something less verbose), and have an intro notebook just going over the basics of 1) integrating arbitrary pytorch Modules with exact GP models and 2) integrating them with variational models (since these differ)? 

The reason I suggest this is that our actual PyTorch integration is actually extremely good beyond just DKL, since we've been careful to make sure everything extends Module and uses Parameter appropriately, and now we have those backward hooks that clear caches. This hypothetical notebook could go over doing things the opposite direction (e.g., `x -> GP -> rsample -> NN`) too, since that has been asked a few times.

**Comment by gpleiss** [comment]
@jacobrgardner yeah that's a good call. In general - I'm also trying to slim down the examples as well to make them less unit-test like. For example - we don't need 4 KISS-GP examples (1D, Kronecker, Additive, 1D Cuda). We can instead have one notebook that has examples of all 4 models.

**Comment by gpleiss** [comment]
Here's a link to the RTD so you can see how the new examples look: https://gpytorch.readthedocs.io/en/refactor_examples/

**Comment by martinjankowiak** [comment]
may i (meekly) suggest you consider adding some sort of CI to your notebooks like we [do](https://github.com/pyro-ppl/pyro/blob/0ce93af91b7129f065676e61548a927d4e0d7878/Makefile#L45) in pyro?

**Comment by gpleiss** [comment]
> may i (meekly) suggest you consider adding some sort of CI to your notebooks like we [do](https://github.com/pyro-ppl/pyro/blob/0ce93af91b7129f065676e61548a927d4e0d7878/Makefile#L45) in pyro?

Done! Thanks for the tip :)

**Comment by gpleiss** [comment]
@jacobrgardner should be ready for review! I recommend checking out https://gpytorch.readthedocs.io/en/refactor_examples/ to see the changes :)

File: examples/02_Scalable_Exact_GPs/index.rst [hunk 1]
```diff
@@ -0,0 +1,102 @@
+Exact GPs with Scalable (GPU) Inference
+========================================
+
+In GPyTorch, Exact GP inference is still our preferred approach to large regression datasets.
+By coupling GPU acceleration with `BlackBox Matrix-Matrix Inference`_ and `LancZos Variance Estimates (LOVE)`_,
```
Comment (jacobrgardner) [inline examples/02_Scalable_Exact_GPs/index.rst]: LanczOs

Also, the link is broken (`pd` -> `pdf`)

**Comment by jacobrgardner** [comment]
Alright, made typo edits + necessary fixes myself in a commit in lieu of giving code review in a pile of notebooks.

**Pull request merged**